src/v/cluster/partition_balancer_planner.h

```diff
@@ -0,0 +1,95 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cluster/health_monitor_types.h"
+#include "cluster/scheduling/partition_allocator.h"
+#include "cluster/topics_frontend.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+
+namespace cluster {
+
+struct ntp_reassignments {
+    model::ntp ntp;
+    allocation_units allocation_units;
+};
+
+struct node_disk_space {
+    model::node_id node_id;
+    double free_space_percent;
+    uint64_t free_space;
+    uint64_t total_space;
+};
+
+struct planner_config {
+    config::binding<double> max_disk_usage;
+    config::binding<size_t> batch_size;
+    config::binding<std::chrono::seconds> node_availability_timeout_sec;
+};
+
+class partition_balancer_planner {
+public:
+    partition_balancer_planner(
+      ss::sharded<topic_table>& topic_table,
+      ss::sharded<partition_allocator>& partition_allocator,
+      planner_config&& config);
+
+    std::vector<ntp_reassignments> get_ntp_reassignments(
+      cluster_health_report&, const std::vector<raft::follower_metrics>&);
+
+private:
+    result<allocation_units> get_reallocation(
+      const partition_assignment& assignments,
+      const topic_metadata& topic_metadata,
+      const std::vector<model::broker_shard>& stable_replicas,
+      const std::vector<model::broker_shard>& old_replicas,
+      size_t partition_size);
+
+    void get_unavailable_nodes_reassigments(std::vector<ntp_reassignments>&);
+
+    void get_full_node_reassigments(std::vector<ntp_reassignments>&);
+
+    void calculate_nodes_with_disk_constraints_violation();
+
+    void
+    calculate_unavailable_nodes(const std::vector<raft::follower_metrics>&);
+
+    std::vector<model::broker_shard>
+    get_stable_replicas(const std::vector<model::broker_shard>& replicas);
+
+    bool is_partition_movement_possible(
+      const std::vector<model::broker_shard>& current_replicas);
+
+    void init_ntp_sizes_and_node_reports_from_health_report(
+      const cluster_health_report& health_report);
+
+    void clear_status();
+
+    size_t
+    get_partition_size(model::ns n, model::topic t, model::partition_id p);
+
+    planner_config _config;
+    ss::sharded<topic_table>& _topic_table;
+    ss::sharded<partition_allocator>& _partition_allocator;
+    uint64_t _cur_batch_size = 0;
+    std::unordered_map<model::ntp, size_t> _ntp_sizes;
+    std::unordered_map<model::node_id, node_health_report> _node_reports;
+    std::unordered_map<model::node_id, uint64_t> _assigned_reallocation_sizes;
+    std::unordered_set<model::node_id> _unavailable_nodes;
+    std::vector<node_disk_space> _nodes_with_disk_constraints_violation;
+    std::unordered_set<model::node_id> _full_nodes;
+    std::set<model::ntp> _moving_partitions;
+    std::unordered_map<model::node_id, uint64_t> _node_released_disk_size;
```

Comment:
This seems like a state of a current reallocation request, maybe it would be nice to pass it in some kind of context object instead of making all of it state of planner ?